### ISSUE TITLE
Polyfill features not adapted to new pageAssets

### ISSUE THREAD

**Issue by craigh** [zikula/core#2629]
Polyfill feature must be implemented

**Comment by shefik** [comment]
@craigh I wrote some documentation to clarify the current use of polyfills in Zikula core.

---

Deploying Polyfills

Firstly, load the following files, immediately after loading jQuery:

`/javascript/js-webshim/dev/polyfiller.js`
`/javascript/js-webshim/dev/polyfiller.init.js`

For best performance, It is recommended that the files are loaded in the <head>, as early as possible.

A new var named “polyfill_features” must be registered by the core:

`PageUtil::registerVar('polyfill_features', true);`

In the plugin “pageaddvar”, if the parameter “name” is set as “javascript”, and the parameter “value” is set as “polyfill”, then the parameter “features” is set as a string to designate the specific polyfill(s) to load. By default, the value is “forms”, since the assumption is that deploying a single polyfill will be a common use case, for use on forms.

For example, this enables native support for HTML5 form validation in browsers that do not fully support it (such as Safari on Mac). In the case of multiple polyfills, separate them by a space (e.g., “forms canvas”). By allowing cross-compatible, native support for HTML5 form validation, other third-party (jQuery) solutions are no longer needed to validate forms.

The parameter “features” is then passed to PageUtil as a multi-value var:

`PageUtil::addVar($name, $value, $features);`

When the function addVar() is executed, the string $features is exploded into an array (using the space as the delimiter). Then, the array is looped, and a new var for “polyfill_features” is set, per $feature:

`PageUtil::addVar('polyfill_features', $feature);`

Other modules may need to further populate the array and insert additional polyfills uniquely. This is why the array was needed for “polyfill_features”.

When JCSSUtil::getJSConfig() is executed, “polyfill_features”  is retrieved:

`$polyfill_features = PageUtil::getVar('polyfill_features');`

Then, the array is imploded again, using a space as a delimiter:

```
if (!empty($polyfill_features)) {
$config['polyfillFeatures'] = implode(' ', $polyfill_features);
}
```

Lastly, the $config['polyfillFeatures'] is JSON-encoded and added to the JavaScript for “Zikula.Config”.

**Comment by craigh** [comment]
1. why not just enable all features always?
2. how is it possible that "other modules may need to further populate the array and insert additional polyfills uniquely"?

**Comment by craigh** [comment]
3. do `html5shiv.js` and/or `modernizr` have anything to do with this?

**Comment by craigh** [comment]
FTR - [html5shiv](https://github.com/aFarkas/html5shiv) "enables use of HTML5 sectioning elements in legacy Internet Explorer and provides basic HTML5 styling for Internet Explorer 6-9, Safari 4.x (and iPhone 3.x), and Firefox 3.x."

**Comment by craigh** [comment]
[modernizr](https://modernizr.com) is "a collection of superfast tests – or “detects” as we like to call them – which run as your web page loads, then you can use the results to tailor the experience to the user."

**Comment by craigh** [comment]
it appears from [these lines of code](https://github.com/zikula/core/blob/1.4/src/javascript/zikula/zikula.core.js#L19-L25) that Modernizr is used to test the browser and load polyfills from `/javascript/polyfills` if required.

@shefik is this legacy and unwanted behavior for Core-2.0? Are the `/javascript/polyfills` an older implementation/solution for the same problem? or do they solve additional problems?

**Comment by shefik** [comment]
@craigh `modernizr` is not required to be loaded separately. However, you can use it with the Webshim polyfills, if you want to do more complex loading of polyfills, based on the results of `modernizr` and condition.

I don't think we need `/javascript/polyfills` anymore. Since, there are already polyfills available more comprehensively in Webshim.

Regarding `html5shiv`, that would need to be loaded, if we are supporting those older browsers.

http://stackoverflow.com/questions/26190945/do-i-need-html5shiv-if-using-webshim-project

**Comment by craigh** [comment]
please:

> why not just enable all features always?
> how is it possible that "other modules may need to further populate the array and insert additional polyfills uniquely"?

**Comment by shefik** [comment]
Don't enable all features, because the library is too large, Only enable what you need. At a minimum, I suggest always enabling `forms`, since that is the common use cause (for form validation).

Other modules can load whatever they need, by just setting the var for polyfill `feature` they desire:

```PageUtil::addVar('polyfill_features', $feature);```

**Comment by craigh** [comment]
> Other modules can load whatever they need, by just setting the var for polyfill feature they desire:

can you provide a use-case where this is necessary or helpful? I am struggling to see it's usefulness, other than setting it when you first declare the need for polyfills.

**Comment by shefik** [comment]
For example, you may desire to also enable forms-ext with forms, which adds support for form widgets (such as date). However, not all forms might have a date field, so it wouldn't be necessary to load the extension all the time.

**Comment by craigh** [comment]
closed by #2634
